### ISSUE TITLE
Eval bug: temp-0 tool-calling is nondeterministic for hybrid Qwen3.6-A3B on Vulkan — flips with prompt-cache mode and across restarts

### ISSUE THREAD

**Issue by jphein** [ggml-org/llama.cpp#26817]
### Name and Version

llama-server, build 221f0f6 (2026-08-02), Vulkan backend
GPU: Intel Arc Pro B60 24GB (Battlemage), Mesa 26.1.6 (VK_NV_cooperative_matrix2 available), kernel 7.0.0-28-generic

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Command line

```
llama-server --model Qwen3.6-35B-A3B-UD-Q4_K_XL.gguf --alias qwen36-coder \
  --device Vulkan0 --port 8091 --host 0.0.0.0 --jinja -fa on \
  --cache-type-k q8_0 --cache-type-v q8_0 --ctx-size 131072 --parallel 2 \
  --kv-unified --n-gpu-layers 999 --threads 8 --batch-size 2048 --ubatch-size 1024
```

### Problem description & steps to reproduce

**With a hybrid-recurrent model (Qwen3.6-35B-A3B, Gated-DeltaNet), reusing cached prompt state can silently drop tool-calling**: at `temperature: 0`, the *identical* OpenAI-compat request emits a well-formed `tool_calls` array when processed with `"cache_prompt": false`, and returns plain prose (no tool call, no error) when cached state is reused. Nothing is logged; the response is otherwise normal — on an agentic lane this silently degrades from "controls things" to "talks about things".

Repro shape (three requests, `temperature 0`, two simple function tools defined; full script available):

1. Fresh server start. Requests with default caching → a request like *"What's the weather in Tokyo right now?"* (expect `get_weather`) returns **no tool_calls** on some post-restart cache states, while *"Turn on the kitchen light"* (expect `control_light`) may or may not fire.
2. Same requests with `"cache_prompt": false` → **always correct** `tool_calls` (dozens of runs, never a failure).
3. After a `cache_prompt:false` pass, default-cache requests *sometimes* recover (the clean full prefill appears to repair the reusable state) — and sometimes don't. Observed both outcomes across several restarts on the same build/flags.

Observed on q8_0 **and** q4_0 KV cache (`-fa on` in both cases), so not KV-quant-specific. The failure correlates with which cached state a request lands on, not with the request content.

This looks related to the hybrid/recurrent checkpointing work (#22384 fixed checkpoint *restore* for DeltaNet/Mamba — this build includes that fix and restore does function; prompt-cache hit rates are otherwise good). Possibly the reused recurrent state is subtly wrong in a way that only shows up as degraded structured-output behavior rather than incoherence. Notably, the same user-visible symptom (prefix-cache hit ⇒ loss of tool calling on Qwen3.6 hybrid) has been reported independently against a different engine (omlx issue #825), which suggests the hybrid-arch state-reuse pattern itself is delicate here.

### First Bad Commit

Unknown — first observed on 221f0f6 in production use; we have no earlier-build A/B for this symptom.

### Relevant log output

```
# temperature 0, tools defined, same server process, minutes apart:

# default cache_prompt:
FAIL weather: NO tool call emitted            # message.content is plain prose
PASS light: control_light({"entity":"kitchen light","state":"on"})

# cache_prompt: false (same requests):
PASS light: control_light({"entity":"kitchen light","state":"on"})
PASS weather: get_weather({"location":"Tokyo"})

# a later restart, after one full clean (cache_prompt:false) pass, back on default cache:
FAIL weather: NO tool call emitted            # "healing" did not persist this time
```

Happy to run instrumented builds, capture `--verbose` slot logs, or bisect prompts further — the setup is a persistent homelab server and the repro harness is scripted.

**Comment by jphein** [comment]
**Correction to my own report after more controlled runs — the causal claim was too strong.**

Further gate runs (same build, flags, prompts, temperature 0) falsify the clean split I reported:

- A later run FAILED the light-control case **with `cache_prompt: false`** (the mode I claimed never fails).
- Minutes later on the same server instance, idle lane: `cache_prompt: false` failed 2 of 2 tool cases while **default caching passed 3/3** — the exact mirror of the original observation.

Revised characterization: **for prompts near the call/don't-call boundary, tool-call emission is nondeterministic at temperature 0 on this stack, in both cache modes, with direction flipping between runs and across restarts.** Cache mode changes *which* perturbation the request sees, but it is not the cause. My current suspicion is numeric jitter from varying batch composition (`--parallel 2` + `--kv-unified`, Vulkan reduction order not batch-invariant) flipping marginal logits — the earlier cache correlation was likely sampling luck across a small number of runs.

Title edited to match. The operational symptom stands (an agentic lane silently alternating between acting and prose), but anyone triaging should treat this as temp-0 nondeterminism with borderline prompts, not a cache-state corruption. Happy to run larger-n sweeps (e.g. 50× per mode per restart) or single-slot (`--parallel 1`) controls if useful — the harness is scripted.
